docs/APP-CONNECTIVITY.md

```diff
@@ -0,0 +1,299 @@
+will reside in the guest (application) domain while the second (the backend in Xen, TAP with
+kvm/qemu) will reside in the host (Dom0 for xen).
+This virtual link is created in a cooperation between EVE and the hypervisor to connect
+application with a [network instance](#network-instances).
+
+By default, VIF is a paravirtualized VirtIO device. A VM application therefore has to have
+a VirtIO driver installed to be able to connect to network instances. Container applications
+are running inside a VM prepared by the EVE, with the driver already included.
+Without HW-assisted virtualization capabilities, EVE will fallback to an emulated e1000 NICs,
+with higher CPU overhead and worse performance. The upside is that the e1000 driver emulates
+a widely used Intel 82540EM network interface card, making it compatible with a broad range
+of operating systems that include built-in support for this hardware.
+
+On the host (EVE) side, VIF has interface name `nbu<vif-num>x<app-num>` and is enslaved
+under the bridge of the corresponding network instance. On the guest side, the VIF
+interface name will be `eth<index>` for a container app, or it will completely depend
+on the OS of a VM application and its interface naming rules.
+
+IP addresses are given to application VIFs using DHCP with a limited lease time. This means
+that a DHCP client should be running inside the guest VM, requesting IP addresses on boot
+and anytime a lease times out. For container apps this is taken care of by the init script
+of the wrapper VM prepared by EVE. When app container starts, virtual interfaces are already
+initialized and have IP addresses assigned.
+VM applications are required to perform DHCP requests themselves. Cloud-init is commonly used
+to instruct the app OS to perform DHCP on selected network interfaces.
+
+Data-plane of VIFs spans across both guest and host network stacks. Packet sent from an application
+is routed by the guest OS (Alpine Linux for container apps) and transmitted across a VIF
+into the host network stack. It arrives via TAP/xen-backend and gets forwarded to the NI bridge.
+Next, ACLs implemented using iptables are applied and the packet is either marked with "allow"
+or "drop" mark. Based on the mark and the src/dst addresses, IP rules either send the packet
+into the dummy blackole interface (dropped), or route the packet according to the NI routing
+table. If the destination is external, packet will be transmitted out through the uplink port.
+If the destination is another app (inside the same NI), packet's dst MAC address will be set
+to the dst app MAC, and it will be forwarded again through the same bridge into the corresponding
+VIF. Finally, the packet is transmitted from the host network stack into the guest network
+stack of the destination application, where it will be processed for local delivery.
+The diagram below depicts all these packet-flow stages:
+
+![packet-flow](./images/eve-app-packet-flow.png)
```

Comment:
Will be good to add anther diagram for app packets inbound (the returned traffic, and also port-mapped packets)

Reply:
Added